perfkitbenchmarker/openstack/os_disk.py

```diff
@@ -0,0 +1,85 @@
+import logging
+import time
+
+from perfkitbenchmarker.openstack import utils as os_utils
+from perfkitbenchmarker import flags
+from perfkitbenchmarker import disk
+from perfkitbenchmarker.openstack import utils
+from perfkitbenchmarker.openstack.utils import retry_authorization
+
+
+FLAGS = flags.FLAGS
+
+
+class OpenStackDisk(disk.BaseDisk):
+    characters = (c for c in "cdefghijklmnopqrstuvwxyz")
+
+    def __init__(self, disk_spec, name, zone, project, image=None):
+        super(OpenStackDisk, self).__init__(disk_spec)
+        self.__nclient = utils.NovaClient(FLAGS.os_auth_url,
+                                          FLAGS.os_tenant,
+                                          FLAGS.os_username,
+                                          FLAGS.os_passwd
+                                          )
+        self.attached_vm_name = None
+        self.attached_vm_id = -1
+        self.image = image
+        self.name = name
+        self.zone = zone
+        self.project = project
+        self.device = ""
+
+    def _Create(self):
+        self._disk = self.__nclient.volumes.create(self.disk_size,
+                                                   display_name=self.name,
+                                                   availability_zone=self.zone,
+                                                   imageRef=self.image,
+                                                   )
+        time.sleep(1)
+        is_unavailable = True
+        while is_unavailable:
+            status = self.__nclient.volumes.get(self._disk.id).status
+            is_unavailable = not (status == "available")
+        self._disk = self.__nclient.volumes.get(self._disk.id)
+
+    @retry_authorization(max_retries=4)
+    def _Delete(self):
+        sleep = 1
+        sleep_count = 0
+        try:
+            self.__nclient.volumes.delete(self._disk)
+            is_deleted = False
+            while not is_deleted:
+                is_deleted = len(self.__nclient.volumes.findall(display_name=self.name)) == 0
+                time.sleep(sleep)
+                sleep_count += 1
+                if sleep_count == 10:
+                    sleep = 5
+        except (os_utils.NotFound, os_utils.BadRequest):
+            logging.info('Volume already deleted')
+
+    def _Exists(self):
+        try:
+            if len(self.__nclient.volumes.findall(display_name=self.name)):
+                return True
+            else:
+                return False
+        except (os_utils.NotFound, os_utils.BadRequest):
+            return False
+
+    def Attach(self, vm):
+        self.attached_vm_name = vm.name
+        self.attached_vm_id = vm.id
+        device_hint_name = "/dev/vd" + self.characters.next()
```

Comment:
Currently, `characters` is a shared class-level variable, equivalent to `OpenStackDisk.characters`. Using `.next()` on it will modify this shared state. If I'm understanding it right, if you have two VMs that each create two OpenStackDisk instances, the first one will use `vdc` and `vdd`, and the second one `vde` and `vdf`. Also, if a benchmark uses a large number of instances, it seems that this could run out of drive letters and throw a *StopIteration* exception.

If you want each VM to have its own independent drive letter namespace, consider following the example in `aws_disk.py`. That also adds locking for thread safety. (Note to self: this would be useful to extract into the base class for reusability.)